test/SelfGuidedFilterTest.cc

```diff
@@ -0,0 +1,218 @@
+/*
+ * Copyright(c) 2019 Netflix, Inc.
+ * SPDX - License - Identifier: BSD - 2 - Clause - Patent
+ */
+
+/******************************************************************************
+ * @file AdaptiveScanTest.cc
+ *
+ * @brief Unit test for self guided filter
+ *
+ * @author Cidana-Wenyao
+ *
+ ******************************************************************************/
+
+#include "gtest/gtest.h"
+#include <stdlib.h>
+// workaround to eliminate the compiling warning on linux
+// The macro will conflict with definition in gtest.h
+#ifdef __USE_GNU
+#undef __USE_GNU  // defined in EbThreads.h
+#endif
+#ifdef _GNU_SOURCE
+#undef _GNU_SOURCE  // defined in EbThreads.h
+#endif
+#include "EbDefinitions.h"
+#include "random.h"
+#include "EbRestoration.h"
+#include "EbUtility.h"
+#include "aom_dsp_rtcd.h"
+#include "util.h"
+
+namespace {
+using ::testing::make_tuple;
+using ::testing::tuple;
+using svt_av1_test_tool::SVTRandom;
+
+using SgrFunc = void (*)(const uint8_t *dat8, int width, int height, int stride,
+                         int eps, const int *xqd, uint8_t *dst8, int dst_stride,
+                         int32_t *tmpbuf, int bit_depth, int highbd);
+
+typedef tuple<SgrFunc, int> FilterTestParam;
+template <typename Sample>
+class SelfGuidedFilterTest : public ::testing::TestWithParam<FilterTestParam> {
+  public:
+    virtual ~SelfGuidedFilterTest() {
+    }
+
+    virtual void SetUp() {
+        input_data_ = (Sample *)aom_memalign(
+            32, in_stride_ * (max_h_ + 32) * sizeof(Sample));
+        output_tst_data_ = (Sample *)aom_memalign(
+            32, out_stride_ * (max_h_ + 32) * sizeof(Sample));
+        output_ref_data_ = (Sample *)aom_memalign(
+            32, out_stride_ * (max_h_ + 32) * sizeof(Sample));
+        tmpbuf_ = (int32_t *)aom_memalign(32, RESTORATION_TMPBUF_SIZE);
+
+        input_ = input_data_ + in_stride_ * 16 + 16;
+        output_tst_ = output_tst_data_ + out_stride_ * 16 + 16;
+        output_ref_ = output_ref_data_ + out_stride_ * 16 + 16;
+        ref_fun_ = apply_selfguided_restoration_c;
+    }
+
+    virtual void TearDown() {
+        aom_free(input_data_);
+        aom_free(output_tst_data_);
+        aom_free(output_ref_data_);
+        aom_free(tmpbuf_);
+        aom_clear_system_state();
+    }
+
+  protected:
+    void RunCorrectnessTest() {
+        tst_fun_ = TEST_GET_PARAM(0);
+        bit_depth_ = TEST_GET_PARAM(1);
+
+        int i, j, k;
+        SVTRandom rnd(bit_depth_, false);
+        // prepare random data
+        for (i = 0; i < NUM_ITERS_; ++i) {
+            for (j = -16; j < max_h_ + 16; ++j)
+                for (k = -16; k < max_w_ + 16; ++k)
+                    input_[j * in_stride_ + k] = rnd.random();
+
+            SVTRandom rnd1(0, SGRPROJ_PRJ_MAX0 + 1 - SGRPROJ_PRJ_MIN0);
```

Comment:
is this test expected same test data for each iterator from random generator? is it better move the random outside the for loop?